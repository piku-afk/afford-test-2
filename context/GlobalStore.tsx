import {
  Dispatch,
  FC,
  PropsWithChildren,
  useContext,
  useReducer,
  createContext,
  useEffect,
} from 'react';
import { initialState } from './initialState';
import { ActionInterface, ActionTypes, reducer } from './reducer';

interface GlobalContextType {
  state: typeof initialState;
  dispatch: Dispatch<ActionInterface>;
}

export const GlobalContext = createContext<GlobalContextType>({
  state: initialState,
  dispatch: () => initialState,
});

export const useGlobalStore = () => useContext(GlobalContext);

type GlobalStoreProps = {};

export const GlobalStore: FC<PropsWithChildren<GlobalStoreProps>> = (props) => {
  const { children } = props;
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    const init = async () => {
      const res = await fetch('http://localhost:5000/filters');
      const data = await res.json();
      const { searchFilter } = data || {};
      const { brands, categories } = searchFilter || {};

      dispatch({ type: ActionTypes.setCategories, payload: categories });
      dispatch({ type: ActionTypes.setBrands, payload: brands });
    };
    init();
  }, []);

  const value = { state, dispatch };

  return (
    <GlobalContext.Provider value={value}>{children}</GlobalContext.Provider>
  );
};
