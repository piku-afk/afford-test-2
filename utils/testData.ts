export const mockCategories = [
  {
    ID: '1',
    name: 'Hello',
    subCategories: [
      { ID: '1', name: 'Piyush', subCategories: [] },
      { ID: '2', name: 'Mahato', subCategories: [] },
    ],
  },
  { ID: '2', name: 'World 2', subCategories: [] },
];

export const mockProducts = [
  {
    id: '1',
    title: 'product 1',
    description: 'some description',
    price: '',
    category: '',
    image: 'http://placeimg.com',
    quantity: 1,
    rating: { rate: '', count: '' },
  },
];

export const mockBrands = [
  { ID: '1', name: 'hello' },
  { ID: '2', name: 'world' },
];
