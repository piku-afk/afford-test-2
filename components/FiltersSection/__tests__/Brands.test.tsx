import * as GlobalContext from '@/context/GlobalStore';
import { mockBrands } from '@/utils/testData';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Brands } from '../brands';

const user = userEvent.setup();
const handleChange = jest.fn();

const MockBrands = () => (
  <Brands value={{ hello: true }} handleChange={handleChange} />
);

describe('Brands', () => {
  beforeEach(() => {
    jest.spyOn(GlobalContext, 'useGlobalStore').mockReturnValue({
      state: { brands: mockBrands, categories: [], cart: [] },
      dispatch: jest.fn,
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<MockBrands />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a checkboxes correctly', () => {
    render(<MockBrands />);

    const brandCheckboxes = screen.getAllByRole('checkbox');
    const helloCheckbox = screen.getByRole('checkbox', { name: /hello/i });
    const worldCheckbox = screen.getByRole('checkbox', { name: /world/i });

    expect(brandCheckboxes).toHaveLength(2);
    expect(helloCheckbox).toBeInTheDocument();
    expect(helloCheckbox).toBeChecked();
    expect(worldCheckbox).toBeInTheDocument();
    expect(worldCheckbox).not.toBeChecked();
  });

  it('renders checkboxes correctly when clicked', async () => {
    render(<MockBrands />);

    const helloCheckbox = screen.getByRole('checkbox', { name: /hello/i });
    const worldCheckbox = screen.getByRole('checkbox', { name: /world/i });

    await user.click(helloCheckbox);
    await user.click(worldCheckbox);

    expect(handleChange).toHaveBeenCalled();
    expect(handleChange).toHaveBeenCalledTimes(2);
  });
});

export {};
