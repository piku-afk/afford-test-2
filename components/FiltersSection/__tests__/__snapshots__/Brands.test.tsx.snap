// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Brands matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-mw5cso-MuiStack-root"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1lt3n9v-MuiFormControlLabel-root"
    >
      <span
        class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root Mui-checked css-usqc7v-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          checked=""
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="CheckBoxIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        hello
      </span>
    </label>
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1lt3n9v-MuiFormControlLabel-root"
    >
      <span
        class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-usqc7v-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        world
      </span>
    </label>
  </div>
</DocumentFragment>
`;
